alpenhorn/Task.py

```diff
@@ -0,0 +1,1021 @@
+""" Classes to run each task in alpenhorn from a queue.
+"""
+
+import datetime as dt
+import logging
+import os
+import re
+import time
+import queue
+import threading
+
+import peewee as pw
+from peewee import fn
+
+from . import acquisition as ac
+from . import archive as ar
+from . import config, db
+from . import update
+from . import storage as st
+from . import util
+
+log = logging.getLogger(__name__)
+
+# Parameters.
+max_time_per_node_operation = 300  # Don't let node operations hog time.
+
+
+class TaskQueue:
+    def add_task(self, task):
+        """Add task to queue."""
+        if self.queue.full():
+            log.warning("Task queue is full ({:d} tasks)".format(self.queue.qsize()))
+        self.queue.put(task)
+
+    def run_tasks(self):
+        """Loop and run tasks from queue."""
+        while True:
+            try:
+                task = self.queue.get()
+            except queue.Empty:
+                pass
+            else:
+                task.run()
+                self.queue.task_done()
+
+    def __init__(self, max_size, num_threads):
+        """Create queue and setup thread pool."""
+        self.queue = queue.Queue(maxsize=max_size)
+
+        threads = []
+        for t in range(num_threads):
+            threads.append(threading.Thread(target=self.run_tasks, daemon=True))
+            threads[t].start()
+
+
+class Task:
+    def __init__(self, node):
+        self.node = node
+
+    def run(self):
+        print("{} run()...".format(type(self).__name__))
+
+
+class IntegrityTask(Task):
+    def run(self):
+        """Check the integrity of file copies on the node."""
+
+        print("{} run() with node: {}".format(type(self).__name__, self.node.name))
+
+        # Find suspect file copies in the database
+        fcopy_query = (
+            ar.ArchiveFileCopy.select()
+            .where(
+                ar.ArchiveFileCopy.node == self.node, ar.ArchiveFileCopy.has_file == "M"
+            )
+            .limit(25)
+        )
+
+        # Loop over these file copies and check their md5sum
+        for fcopy in fcopy_query:
+            fullpath = "%s/%s/%s" % (
+                self.node.root,
+                fcopy.file.acq.name,
+                fcopy.file.name,
+            )
+            log.info('Checking file "%s" on node "%s".' % (fullpath, self.node.name))
+
+            # If the file exists calculate its md5sum and check against the DB
+            if os.path.exists(fullpath):
+                if util.md5sum_file(fullpath) == fcopy.file.md5sum:
+                    log.info("File is A-OK!")
+                    fcopy.has_file = "Y"
+                    copy_size_b = os.stat(fullpath).st_blocks * 512
+                    fcopy.size_b = copy_size_b
+                else:
+                    log.error("File is corrupted!")
+                    fcopy.has_file = "X"
+            else:
+                log.error("File does not exist!")
```

Comment:
Let's document the fullpaths in these error logs. Because now there might be multiple threads going, and the logs might get interleaved.